Add RAG chatbot tutorial

arcanaxion opened the pull request:
Tutorial for the demo hosted at https://github.com/Avaiga/demo-gpt-4o/tree/rag.

Screenshots:
[rag-chatbot-docs-screenshots.zip](https://github.com/user-attachments/files/16159376/rag-chatbot-docs-screenshots.zip)

arcanaxion committed:
Add RAG chatbot tutorial

AlexandreSajus reviewed (changes requested):
Great tutorial! Nothing to say about the content itself.

Please make sure the installation works:
- currently I'm stuck at the git clone with this error:
```bash
PS C:\Users\asaju\Desktop> git clone -b rag --single-branch git@github.com:Avaiga/demo-gpt-4o.git
Cloning into 'demo-gpt-4o'...
git@github.com: Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```
- Good idea on using Docker to install the weird dependencies. Will `docker compose up` work without a Dockerfile in the `demo-gpt-4o` repo?

Also don't forget to add this tutorial to `mkdocs.yml_template`

FlorianJacta commented:
Could you build the doc and show us the different pages? This tutorial should change by using the chat visual element when this will be used in the original demo.

FlorianJacta commented on docs/tutorials/fundamentals/5_rag_chatbot/index.md:
````diff
@@ -0,0 +1,306 @@
+    state.query_message = ""
+```
+
+This function calls the *create_conv* function **twice**: First to add the **user’s message** to the chat UI, then a
````
*user’s* is using a bad formatted character. This will create issues.

arcanaxion committed:
Change single quote to apostrophe

arcanaxion committed:
Add tutorial to mkdocs.yml_template

arcanaxion commented:
Thanks!

> Please make sure the installation works:

Ah, that would only work for users that have configured SSH. I'll change it to use HTTPS which would be more common.

> Good idea on using Docker to install the weird dependencies. Will `docker compose up` work without a Dockerfile in the `demo-gpt-4o` repo?

No, it wouldn't work for the develop branch. The previous command would clone the rag branch though, which does have the Dockerfile.

> Also don't forget to add this tutorial to `mkdocs.yml_template`

Done!

arcanaxion commented:
> Could you build the doc and show us the different pages? 

Added a zip file with the screenshots to the main comment.

> This tutorial should change by using the chat visual element when this will be used in the original demo.

Should I change it now or after it's released on stable?

arcanaxion committed:
Change git clone url from ssh to https

FlorianJacta commented:
Did you create the screenshots before putting the correct line in the *mkdocs.yml_template*? Your article does not have navigation on the left, which indicates that it was not correctly built.

I suspect it was fixed when you changed the *mkdocs.yml_template*. Could you check?

Also, to make your life easier, you can save a page with "ctrl+s" on your article as a whole file. Then, you can compress it as a zip and drop it in the PR. You don't have to do it now, as you created the proper screenshots.

arcanaxion commented:
> Did you create the screenshots before putting the correct line in the _mkdocs.yml_template_? Your article does not have navigation on the left, which indicates that it was not correctly built.
> 
> I suspect it was fixed when you changed the _mkdocs.yml_template_. Could you check?

It is indeed present after the changes to mkdocs.yml_template

> Also, to make your life easier, you can save a page with "ctrl+s" on your article as a whole file. Then, you can compress it as a zip and drop it in the PR. You don't have to do it now, as you created the proper screenshots.

Great, thanks 😄

FlorianJacta commented:
> Should I change it now or after it's released on stable?

I believe @AlexandreSajus should change it on all demos around LLM at some point.

AlexandreSajus reviewed (changes requested):
The docker install does not work on my side. Please ensure you add a step in the installation part to create a .env file with the Hugging Face API key and that the installation works without error. Otherwise, you get this error:
```bash
 => ERROR [app 10/10] COPY .env main.py main.css ./                                              0.0s 
------
 > [app 10/10] COPY .env main.py main.css ./:
------
failed to solve: failed to compute cache key: failed to calculate checksum of ref d672100b-6a47-4c33-9d64-760ab9839d0d::fgmb97wefd84r1hkvbd3k0qp1: "/.env": not found
```

Even after fixing this, I get this error:
```bash
 => ERROR [app  5/10] RUN xargs apt install -y < packages.txt                                           0.9s 
------
 > [app  5/10] RUN xargs apt install -y < packages.txt:
0.389
0.389 WARNING: apt does not have a stable CLI interface. Use with caution in scripts.
0.389
0.391 Reading package lists...
0.772 Building dependency tree...
0.879 Reading state information...
0.885 E: Unable to locate package tesseract-ocr
0.885 E: Unable to locate package libtesseract-dev
0.885 E: Unable to locate package libpoppler-dev
0.885 E: Unable to locate package poppler-utils
0.885 E: Unable to locate package libgl1
------
failed to solve: process "/bin/sh -c xargs apt install -y < packages.txt" did not complete successfully: exit code: 123
```

arcanaxion commented:
> Please ensure you add a step in the installation part to create a .env file with the Hugging Face API key and that the installation works without error.

My mistake, I didn't catch this because I forgot to remove the old image when testing the installation steps. Adding this step now.

> Even after fixing this, I get this error

I'm not getting this error. Can you try `docker compose up --build --force-recreate`? I wonder if it's caused by your previous failed build

arcanaxion committed:
Add installation step for creating .env file

AlexandreSajus commented:
> force

I still get an error with that command. Do you have another PC on which to test the installation? If not, I'll try on my desktop when I get home

arcanaxion commented:
> I still get an error with that command. Do you have another PC on which to test the installation? If not, I'll try on my desktop when I get home

I reproduced it on Docker on Windows. Fixed in https://github.com/Avaiga/demo-gpt-4o/pull/2

AlexandreSajus committed:
Adds a step for the user to add his own pdfs

AlexandreSajus reviewed (approved):
Good find! So, what is the fix exactly? You need to apt-get update and install packages within the same command line, or it fails?

You left `COPY packages.txt`, which caused an error, so I removed that step, and now it works perfectly. I also added an optional step in the installation process for the user to add custom PDFs to the `pdfs` folder if he wants to.

Approved!

arcanaxion commented:
> Good find! So, what is the fix exactly? You need to apt-get update and install packages within the same command line, or it fails?

Maybe a caching issue between the `apt-get upgrade` and `xargs` layers, but then `--force-recreate` should have resolved that. Even stranger that it works for me on Linux but not on Windows. I'll investigate further another time but for now it's following best practices and it works so it should be fine.

> You left `COPY packages.txt`, which caused an error, so I removed that step, and now it works perfectly. I also added an optional step in the installation process for the user to add custom PDFs to the `pdfs` folder if he wants to.

My mistake 😬. Thanks for the changes

> Approved!

Great, please merge it when ready since I don't have permissions
